### DISCUSSION TITLE
how to set color of progress bar?

### DISCUSSION THREAD

**Discussion by m4b** [hecrj/iced#1503] [Q&A] [answered]
In previous versions of iced, setting the color of a progress bar was easily done, by calling `.style` method on the progress bar widget, and implementing one's own "style struct".
With the new style changes in iced, I do not see a way to programmatically change the color of a progress bar (without afaics, implementing an entire theme for oneself), which seems like a regression.
There is a `Custom(fn(&Theme) -> progress_bar::Appearance),` variant, but because this is a function pointer, one cannot pass in programmatic values computed into a closure to modify the returned appearance.
It's possible I'm missing something, but today with the new style system, I do not see how a user can change the color of a progress bar without implementing an entire custom theme.

**Comment by arjunair86** [comment]
I think this is a known behaviour: https://github.com/iced-rs/rfcs/blob/master/text/0006-theming.md
>An important drawback is that the current design does not allow for widget customization outside of a Theme. For instance, if a user is using an existing Theme, they will not be able to easily fine-tune the specific properties of the styling of a Button to their liking. They will be forced to create a new custom theme.

**Comment by m4b** [comment, marked as answer]
Ok so as of the recent changes, which are in 0.5, because a `Box<dyn` is now there, instead of a function pointer, we can change colors like this without using a wholesale theme, which is great!

https://docs.rs/iced_style/latest/iced_style/theme/enum.ProgressBar.html#variant.Custom

Something like this is possible (adapted from the 0.4 and lesser mod based styles, and not guaranteed to compile as is, just an example):

```rust
mod style {
    use super::*;
    use iced::progress_bar::StyleSheet;
    use iced_style::progress_bar;

    pub struct ProgressBar(pub(super) [f32; 3]);

    impl StyleSheet for ProgressBar {
        type Style = iced_style::Theme;

        fn appearance(&self, _style: &Self::Style) -> progress_bar::Appearance {
            progress_bar::Appearance {
                background: Background::Color(Color::from_rgb(0.6, 0.6, 0.6)),
                bar: Background::Color(Color::from(self.0)),
                border_radius: 10.0,
            }
        }
    }

    impl From<ProgressBar> for iced_style::theme::ProgressBar {
        fn from(bar: ProgressBar) -> Self {
            iced::theme::ProgressBar::Custom(Box::new(bar))
        }
    }
}
```

So I think i can mark this as answered
